Fixes #3168 Test Sending Email Issues

thabaum opened the pull request:
<!-- 
  Please read contribution guideline first: https://github.com/dnnsoftware/Dnn.Platform/blob/development/CONTRIBUTING.md 
-->

<!-- 
  Please make sure that there is a corresponding issue created and reference it in the PR by writing
  `Fixes #123` or `Closes #123`. 
  A PR without an accompanying issue will be accepted and merged on a very rare occasion
-->
Fixes
https://github.com/dnnsoftware/Dnn.Platform/issues/3168
https://www.dnnsoftware.com/forums/threadid/86505/scope/posts/change-portal-administrator-email-address
## Summary
<!-- 
  Please describe the code changes as you see fit so that the reviewers have an easier task understanding what changed and why.
  New unit tests will be highly appreciated.
-->
I may get shot down here but it is worth a try.   Thanks for the dnndocs.com it helped me find some things I was looking for. I found a typo there I will submit a fix for as well.

Please help me if I went wrong here it just didnt make sense the way it was written when setting the To and From addresses nothing referenced the portal administrator email.  And the send to address it feels to me like it should be going to whoever is logged in testing it so they can see it works.

I also am not sure if this actually uses the email being set in Host SMTP Settings.  For testing the emails this is where it needs set so any changes to be made...

bdukes commented:
@thabaum I've rebased your branch on `release/9.4.x`

thabaum commented:
sorry, trying to figure this out here not trying to get put anywhere special I need to understand this a bit more please direct, that is strike 3 however, that is my next goal... figure out what I am doing wrong with the PR's I think I need more coffee, time for a brake :) I will get back to work on this.  I want to get one right.  I will hold off on my next one until I re-review all the latest tutorials.

Now when you say Rebase, can you link me to how 

> 
> 
> @thabaum I've rebased your branch on `release/9.4.x`

ok now how did you do that :)

thabaum commented:
OK i see the changes no commits now.

bdukes commented:
When you make your changes, you want to start on the `release/9.4.x` branch (i.e. that would be the starting point of your new branch for the PR).  Since you started on `development`, I did something like `git rebase origin/release/9.4.x --interactive`, which replays all of the commits from your branch onto `release/9.4.x`.  It does this _interactively_, meaning that it opens up a text file with a list of the commits it is going to replay, and allows you to make adjustments to it.  In this case, I adjusted that list by deleting all but your three commits.

bdukes commented:
[Build error](https://dev.azure.com/dotnet/DNN/_build/results?buildId=24314&view=logs&j=fd490c07-0b22-5182-fac9-6d67fe1e939b&t=a04569be-1b08-5c9a-7e65-a83076350a0c):
>`Services\ServerSettingsSmtpAdminController.cs`(123,64): error CS0103: The name `Portals` does not exist in the current context [`D:\a\1\s\Dnn.AdminExperience\Extensions\Content\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj`]

thabaum commented:
I think I have addressed all concerns here.

If all is correct the Send From email address if smtpHostMode is TRUE then host email and if false portal administrator email is used.  

The address it will send to will use the users email address that is currently used to run the test.  You can then switch your administrator email account to whatever you like and test emails going to msn, yahoo, google... with ease I believe if this works as planned.

thabaum commented:
> When you make your changes, you want to start on the `release/9.4.x` branch (i.e. that would be the starting point of your new branch for the PR). Since you started on `development`, I did something like `git rebase origin/release/9.4.x --interactive`, which replays all of the commits from your branch onto `release/9.4.x`. It does this _interactively_, meaning that it opens up a text file with a list of the commits it is going to replay, and allows you to make adjustments to it. In this case, I adjusted that list by deleting all but your three commits.


> 
> 
> @thabaum please rebase this PR on the current `release/9.4.x` branch. Right now you have 44 commits in this PR (many from other PRs). Thanks 🙏🏼

I believe I know where I am going wrong as I need to start out with the branch I am working on, it's never that easy is it :).  I will hopefully get it down for the next PR sorry it is taking me a few times to really understand what I am doing here. Thanks you for the patience and support.

thabaum commented on Dnn.AdminExperience/Extensions/Content/Dnn.PersonaBar.Extensions/Services/ServerSettingsSmtpAdminController.cs:
```diff
@@ -120,8 +120,8 @@ public HttpResponseMessage SendTestEmail(SendTestEmailRequest request)
-                var mailFrom = Host.HostEmail;
-                var mailTo = smtpHostMode ? Host.HostEmail : PortalSettings.UserInfo.Email;
+                var mailFrom = smtpHostMode ? Host.HostEmail : PortalSettings.Email;
+                var mailTo = PortalSettings.Current.UserInfo.Email;
```
var mailFrom = smtpHostMode ? Host.HostEmail : PortalSettings.Email;

If I am reading this statement correctly:

If smtHostMode is "=" TRUE then use host email alternatively ":" use use Portal Admin Email

The statement above this sets smtpHostMode will be true when smtpHostMode == h , otherwise it will be false.

var smtpHostMode = request.SmtpServerMode == "h";

"==" should return a true/false boolean datatype value.

so if SmtpServerMode is false it should use the alternate address

A really slick if statement packed in one line.

thabaum commented:
A few issues should reference:

https://github.com/dnnsoftware/Dnn.Platform/issues/2298
https://github.com/dnnsoftware/Dnn.Platform/issues/2227
https://github.com/dnnsoftware/Dnn.Platform/issues/2298
https://github.com/dnnsoftware/Dnn.Platform/issues/3168
and the ones dating back to DNN version 4 like this one in 2006
https://www.dnnsoftware.com/forums/threadid/86505/scope/posts/change-portal-administrator-email-address

I will now verify these portal settings are used for actually sending the email notifications.

It feels like if I introduce a section of an issue and not combine a huge issue it is easier to chew on.  Similar to another issue created that I want to be able to edit site settings without having to log into each portal.  Take on the entire project I think is a lot for one person to do, however take on one page/PR at a time and it will get done without feeling so overwhelmed.

I am also interested in trying to work on some bigger issues with a few takers that can help collaborate on an issue together.  I just need some help getting started with a few things.

I want to add two way verification, SMS and another security feature to login along with fix any registration issues as needed so the entire system that has to do with user communication and account security is 100% by version 10.

This is my own personal goal if anyone is taking on these tasks please let me know so I can push on some other direction.

thabaum commented:
@david-poindexter Everything good here?

DHCuthill commented:
This condition of a placeholder default From Address **(admin@changeme.invalid)** being sent in system emails did not occur in version 
DNN PLATFORM v. 09.06.02 (0)
It has come back in the latest version 
DNN PLATFORM v. 09.09.00 (0)

The error has returned.  This makes the portal unfit for release to my customers.
Any advice on how to deal with this would be welcome.

DHCuthill commented:
I have a solution that will do for me.
1) Add a new user "Administrator" and ensure there is a valid email address
2) Add the new user "Administrator" to the Role of "Administrators
3) Go to Login Settings/Basic Login Settings/ Primary Administrator
4) From the drop-down-list select the new user you created (Administrator)
5) To test the site auto-email function create a test user and check that the email came from the "Administrator" email address

valadas commented:
That default email with .invalid domain name was selected by design so it can never be registered by anyone for security reasons, so as part of portal/site setup one needs to replace it by a valid email.
